lib/services/v2.0/kc2_service.dart

```diff
@@ -1,473 +1,1375 @@
+
+        final transactionEvents = block.events
+            .where((event) => event.extrinsicIndex == transactionIndex)
+            .toList();
+
+        try {
+          _processTransaction(
+              hash: hash,
+              userInfo: userInfo,
+              tx: transaction,
+              txEvents: transactionEvents,
+              timestamp: timestamp,
+              blockNumber: blockNumber,
+              blockIndex: transactionIndex);
+        } catch (e) {
+          debugPrint('>>> failed block tx: $e');
+        }
+      });
+
+      return blockNumber;
+    } catch (e) {
+      debugPrint('Failed to process block: $e');
       rethrow;
     }
   }
 
-  /// Fetch list who participate in karma rewards distribution
-  Future<List<KC2UserInfo>> getLeaderBoard() async {
-    try {
-      List<dynamic> result = await callRpc('community_getLeaderBoard', []);
-      return result.map((e) => KC2UserInfo.fromJson(e)).toList();
-    } on PlatformException catch (e) {
-      debugPrint('Failed to get leader board: ${e.message}');
-      rethrow;
-    }
+  Map<String, dynamic> _decodeTransaction(Input input) {
+    return ExtrinsicsCodec(chainInfo: chainInfo).decode(input);
   }
 
-  /// Fetch information about chain like last block time, rewards, etc
-  Future<BlockchainStats> getBlockchainStats() async {
+  ///
+  /// Process a single kc2 tx
+  Future<void> _processTransaction({
+    // local user we are processing this tx for
+    required KC2UserInfo userInfo,
+    required Map<String, dynamic> tx,
+    required List<KC2Event> txEvents,
+    required int timestamp,
+    required String? hash,
+    required BigInt blockNumber,
+    required int blockIndex,
+  }) async {
     try {
-      Map<String, dynamic> result =
-          await callRpc('chain_getBlockchainData', []);
-      return BlockchainStats.fromJson(result);
+      final String pallet = tx['calls'].key;
+      final String method = tx['calls'].value.key;
+      final args = tx['calls'].value.value;
+
+      final String? signer = _getTransactionSigner(tx);
+
+      if (signer == null) {
+        debugPrint(">>> skipping unsigned tx $pallet/$method");
+        return;
+      }
+
+      /// Use provided hash or generate one if needed
+      hash ??=
+          '0x${hex.encode(Hasher.blake2b256.hash(ExtrinsicsCodec(chainInfo: chainInfo).encode(tx)))}';
+
+      debugPrint("Processing tx $pallet/$method. txHash: $hash");
+
+      final failedEventData = txEvents
+          .where((event) => event.eventName == 'ExtrinsicFailed')
+          .firstOrNull
+          ?.data;
+
+      final ChainError? chanError = _getChainError(failedEventData);
+
+      if (pallet == 'Identity' &&
+          method == 'new_user' &&
+          newUserCallback != null) {
+        final txAccountId = encodeAccountId(args['account_id'].cast<int>());
+        // @HolyGrease - shouldn't we just check signer here?
+        if (signer == userInfo.accountId || txAccountId == userInfo.accountId) {
```

Comment:
Yes you right we can check only for signer here
```suggestion
        if (signer == userInfo.accountId) {
```